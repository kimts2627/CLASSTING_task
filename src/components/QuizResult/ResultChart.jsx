import Chart from "react-google-charts";
import Spinner from "react-bootstrap/Spinner";
import { useEffect, useState } from "react";
import { useDispatch } from "react-redux";
import { resetResult } from "../../modules/results";
import { resetQuiz } from "../../modules/quiz";

const ResultChart = ({ correct, incorrect, quizNum }) => {
  const [result, setResult] = useState("");

  const dispatch = useDispatch();

  const onResetResult = () => {
    dispatch(resetResult());
  };

  const onResetQuiz = () => {
    dispatch(resetQuiz());
  };

  useEffect(() => {
    if (correct === quizNum) {
      setResult("Perfect!!");
    } else if (correct / quizNum >= 0.5) {
      setResult("Not Bad!");
    } else if (correct !== 0) {
      setResult("Try Again..");
    } else {
      setResult("Fail...");
    }
  }, []);

  const restartQuiz = () => {
    onResetQuiz();
    onResetResult();
  };

  return (
    <>
      <Chart
        width={"350px"}
        height={"350px"}
        chartType="PieChart"
        loader={<Spinner animation="border" />}
        data={[
          ["chart", "number"],
          ["Correct", correct],
          ["Incorrect", incorrect],
        ]}
        options={{
          legend: "none",
          pieSliceText: "label",
          pieStartAngle: 100,
          backgroundColor: "none",
        }}
        rootProps={{ "data-testid": "2" }}
      />
      <article className="w-full h-1/3 flex flex-col items-center justify-between">
        <h1
          className={`text-4xl Kanit ${
            result === "Perfect!!" || result === "Not Bad!" ? "text-green-500" : "text-red-700"
          }`}
        >
          {result}
        </h1>
        <p className="Kanit -mt-20">{`You got ${correct} out of ${quizNum} questions`}</p>
        <button className="w-20 bg-white opacity-70 active:opacity-90 rounded-md" onClick={restartQuiz}>
          Restart
        </button>
      </article>
    </>
  );
};

export default ResultChart;
