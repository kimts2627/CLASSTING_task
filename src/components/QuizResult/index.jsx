import React, { useEffect, useRef, useState } from "react";
import { useSelector } from "react-redux";
import ResultChart from "./ResultChart";
import Spinner from "react-bootstrap/Spinner";
import Timer from "../QuizPage/Timer";

const QuizResult = () => {
  const quiz = useSelector((state) => state.quiz.quiz);
  const result = useSelector((state) => state.results.quizResult);
  const [correct, setCorrect] = useState(0);
  const [incorrect, setIncorrect] = useState(0);

  const wrapRef = useRef();

  useEffect(() => {
    wrapRef.current.style.backgroundColor = "rgba(255, 255, 255, 0.6)";
  }, []);

  useEffect(() => {
    let o = 0;
    let x = 0;
    for (let i = 0; i < result.length; i++) {
      if (result[i] === quiz[i].correct_answer) {
        o++;
      } else {
        x++;
      }
    }
    setCorrect(o);
    setIncorrect(x);
  }, []);
  return (
    <section className="w-4/5 h-4/5 mt-20 rounded-xl flex flex-col items-center justify-evenly" ref={wrapRef}>
      {correct + incorrect === quiz.length ? (
        <>
          <Timer status="resultpage" />
          <ResultChart correct={correct} incorrect={incorrect} quizNum={quiz.length} />
        </>
      ) : (
        <Spinner animation="border" />
      )}
    </section>
  );
};

export default QuizResult;
