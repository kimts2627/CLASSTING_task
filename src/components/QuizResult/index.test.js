import React from "react";
import { mount, configure } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import QuizResult from "./index";
import ResultChart from "./ResultChart";
import { store } from "../../index";
import { Provider } from "react-redux";

configure({ adapter: new Adapter() });
let component = null;

describe("QuizResult", () => {
  describe("QuizResult", () => {
    it("컴포넌트가 정상적으로 렌더링 되어야 함", () => {
      component = mount(
        <Provider store={store}>
          <QuizResult />
        </Provider>
      );
      expect(component.length).toBe(1);
    });
  });

  describe("ResultChart", () => {
    it("컴포넌트가 정상적으로 렌더링 되어야 함", () => {
      component = mount(
        <Provider store={store}>
          <ResultChart correct={3} incorrect={7} quizNum={10} />
        </Provider>
      );
      expect(component.length).toBe(1);
    });

    it("10개중 10개를 맞췄을 경우 h1 tag에 Perfect!!를 프린트 해야 함", () => {
      component = mount(
        <Provider store={store}>
          <ResultChart correct={10} incorrect={0} quizNum={10} />
        </Provider>
      );
      let resultText = component.find("h1").text();
      expect(resultText).toEqual("Perfect!!");
    });

    it("10개중 0개를 맞췄을 경우 h1 tag에 Fail...을 프린트 해야 함", () => {
      component = mount(
        <Provider store={store}>
          <ResultChart correct={0} incorrect={10} quizNum={10} />
        </Provider>
      );
      let resultText = component.find("h1").text();
      expect(resultText).toEqual("Fail...");
    });

    it("5개중 3개를 맞췄을 경우 h1 tag에 Not Bad!를 프린트 해야 함", () => {
      component = mount(
        <Provider store={store}>
          <ResultChart correct={3} incorrect={2} quizNum={5} />
        </Provider>
      );
      let resultText = component.find("h1").text();
      expect(resultText).toEqual("Not Bad!");
    });

    it("5개중 3개를 맞췄을 경우 p tag에 결과를 프린트 해야 함", () => {
      component = mount(
        <Provider store={store}>
          <ResultChart correct={3} incorrect={2} quizNum={5} />
        </Provider>
      );
      let resultText = component.find("p").text();
      expect(resultText).toEqual("You got 3 out of 5 questions");
    });
  });

  it("스냅샷과 결과물이 일치해야 함", () => {
    expect(component).toMatchSnapshot();
  });
});
