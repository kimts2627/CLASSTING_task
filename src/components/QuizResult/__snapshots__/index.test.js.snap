// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizResult 스냅샷과 결과물이 일치해야 함 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ResultChart
    correct={3}
    incorrect={2}
    quizNum={5}
  >
    <div
      className="w-full h-full flex flex-col items-center justify-evenly"
    >
      <Chart
        chartActions={null}
        chartEvents={null}
        chartType="PieChart"
        chartWrapperParams={Object {}}
        className=""
        columns={null}
        controls={null}
        data={
          Array [
            Array [
              "chart",
              "number",
            ],
            Array [
              "Correct",
              3,
            ],
            Array [
              "Incorrect",
              2,
            ],
          ]
        }
        diffdata={null}
        formatters={null}
        getChartEditor={null}
        getChartWrapper={[Function]}
        graphID={null}
        graph_id={null}
        height="350px"
        legendToggle={false}
        legend_toggle={false}
        loader={
          <Spinner
            animation="border"
          />
        }
        options={
          Object {
            "backgroundColor": "none",
            "legend": "none",
            "pieSliceText": "label",
            "pieStartAngle": 100,
          }
        }
        render={null}
        rootProps={
          Object {
            "data-testid": "2",
          }
        }
        rows={null}
        spreadSheetQueryParameters={
          Object {
            "gid": 1,
            "headers": 1,
          }
        }
        spreadSheetUrl={null}
        style={Object {}}
        toolbarID={null}
        toolbarItems={null}
        width="350px"
      >
        <ContextProvider
          value={
            Object {
              "chartActions": null,
              "chartEvents": null,
              "chartType": "PieChart",
              "chartWrapperParams": Object {},
              "className": "",
              "columns": null,
              "controls": null,
              "data": Array [
                Array [
                  "chart",
                  "number",
                ],
                Array [
                  "Correct",
                  3,
                ],
                Array [
                  "Incorrect",
                  2,
                ],
              ],
              "diffdata": null,
              "formatters": null,
              "getChartEditor": null,
              "getChartWrapper": [Function],
              "graphID": null,
              "graph_id": null,
              "height": "350px",
              "legendToggle": false,
              "legend_toggle": false,
              "loader": <Spinner
                animation="border"
              />,
              "options": Object {
                "backgroundColor": "none",
                "legend": "none",
                "pieSliceText": "label",
                "pieStartAngle": 100,
              },
              "render": null,
              "rootProps": Object {
                "data-testid": "2",
              },
              "rows": null,
              "spreadSheetQueryParameters": Object {
                "gid": 1,
                "headers": 1,
              },
              "spreadSheetUrl": null,
              "style": Object {},
              "toolbarID": null,
              "toolbarItems": null,
              "width": "350px",
            }
          }
        >
          <Spinner
            animation="border"
          >
            <div
              className="spinner-border"
            />
          </Spinner>
          <GoogleChartLoader
            onError={[Function]}
            onLoad={[Function]}
          >
            <Script
              attributes={Object {}}
              onCreate={[Function]}
              onError={[Function]}
              onLoad={[Function]}
              url="https://www.gstatic.com/charts/loader.js"
            />
          </GoogleChartLoader>
        </ContextProvider>
      </Chart>
      <article
        className="w-full h-1/4 flex flex-col items-center justify-between -mt-20"
      >
        <h1
          className="text-4xl Kanit text-green-500"
        >
          Not Bad!
        </h1>
        <p
          className="Kanit -mt-10"
        >
          You got 3 out of 5 questions
        </p>
        <button
          className="w-20 bg-white opacity-70 active:opacity-90 rounded-md"
          onClick={[Function]}
        >
          Restart
        </button>
      </article>
    </div>
  </ResultChart>
</Provider>
`;
